onnxruntime/core/providers/cpu/tensor/where_op.cc

```diff
@@ -0,0 +1,195 @@
+    if (scalar_value != T{}) {
+      output = EigenVectorMap<T>::PlainObject::Constant(vector_value.size(), scalar_value);
+    } else {
+      output = vector_value;
+    }
+  };
+
+  BroadcastLoop(
+      *merge_broadcaster, *merge_broadcast_output,
+      [merge_scalar_vector](EigenVectorMap<T> output, const T& X_selection, ConstEigenVectorMap<T> Y_selection) {
+        merge_scalar_vector(output, X_selection, Y_selection);
+      },
+      [merge_scalar_vector](EigenVectorMap<T> output, ConstEigenVectorMap<T> X_selection, const T& Y_selection) {
+        merge_scalar_vector(output, Y_selection, X_selection);
+      },
+      [](EigenVectorMap<T> output, ConstEigenVectorMap<T> X_selection, ConstEigenVectorMap<T> Y_selection) {
+        output = X_selection.binaryExpr(Y_selection, [](T x, T y) -> T {
+          return x != T{} ? x : y;
+        });
+      });
+}
+
+template <typename T>
+std::enable_if_t<!IsEigenScalarCompatible<T>, void>
+MergeBroadcastLoop(TBroadcaster<T, T>* merge_broadcaster, TBroadcastOutput<T>* merge_broadcast_output) {
+  const auto merge_scalar_vector = [](gsl::span<T> output, const T& scalar_value, gsl::span<const T> vector_value) {
+    if (scalar_value != T{}) {
+      std::fill(output.begin(), output.end(), scalar_value);
+    } else {
+      std::copy(vector_value.cbegin(), vector_value.cend(), output.begin());
+    }
+  };
+
+  BroadcastLoopSpan(
+      *merge_broadcaster, *merge_broadcast_output,
+      [merge_scalar_vector](gsl::span<T> output, const T& X_selection, gsl::span<const T> Y_selection) {
+        merge_scalar_vector(output, X_selection, Y_selection);
+      },
+      [merge_scalar_vector](gsl::span<T> output, gsl::span<const T> X_selection, const T& Y_selection) {
+        merge_scalar_vector(output, Y_selection, X_selection);
+      },
+      [](gsl::span<T> output, gsl::span<const T> X_selection, gsl::span<const T> Y_selection) {
+        std::transform(
+            X_selection.cbegin(), X_selection.cend(), Y_selection.cbegin(), output.begin(),
+            [](const T& x, const T& y) {
+              return x != T{} ? x : y;
+            });
+      });
+}
+}  // namespace
+
+template <typename T>
+Status Where<T>::Compute(OpKernelContext* context) const {
+  const Tensor* const condition = context->Input<Tensor>(0);
+  const Tensor* const X = context->Input<Tensor>(1);
+  const Tensor* const Y = context->Input<Tensor>(2);
+  ORT_ENFORCE(condition && X && Y, "condition, X, and Y inputs are required!");
+
+  // The current implementation is limited to broadcasting over two tensors at once.
+  // So, we first broadcast over condition and X to select the values from X:
+  //   X_selection = condition ? X : default value
+  // Similarly, we broadcast over condition and Y to select the values from Y:
+  //   Y_selection = !condition ? Y : default value
+  // Finally, we broadcast over and merge X_selection and Y_selection:
+  //   output = (X_selection != default value) ? X_selection : Y_selection
```

Comment:
>X_selection != default value [](start = 17, length = 28)

Is it possible for the selected value to be == to the default value for T? If so do you accidentally chose the wrong one?

Reply:
If the desired value (say, from X_selection) happens to be the default value, then the alternative (from Y_selection) cannot be something other than the default value. So in that case, the result is the default value.